hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java

```diff
@@ -0,0 +1,694 @@
+   * @return true if the key is created, false otherwise
+   */
+  @Override
+  public boolean createDirectory(String pathStr) throws IOException {
+    LOG.trace("creating dir for path: {}", pathStr);
+    incrementCounter(Statistic.OBJECTS_CREATED);
+    OFSPath ofsPath = new OFSPath(pathStr);
+    // Volume name unspecified, return failure
+    if (ofsPath.getVolumeName().length() == 0) {
+      return false;
+    }
+    // Handle where only volume is specified in pathStr
+    if (ofsPath.getBucketName().length() == 0) {
+      objectStore.createVolume(ofsPath.getVolumeName());
+      return true;
+    }
+    String keyStr = ofsPath.getKeyName();
+    try {
+      OzoneBucket bucket = getBucket(ofsPath, true);
+      // if keyStr is empty, it indicates that only volume or volume+bucket is
+      // given in pathStr, so getBucket() above should've handled the creation
+      // of volume/bucket already.
+      if (keyStr != null && keyStr.length() > 0) {
+        bucket.createDirectory(keyStr);
+      }
+    } catch (OMException e) {
+      if (e.getResult() == OMException.ResultCodes.FILE_ALREADY_EXISTS) {
+        throw new FileAlreadyExistsException(e.getMessage());
+      }
+      throw e;
+    }
+    return true;
+  }
+
+  /**
+   * Helper method to delete an object specified by key name in bucket.
+   *
+   * @param path path to a key to be deleted
+   * @return true if the key is deleted, false otherwise
+   */
+  @Override
+  public boolean deleteObject(String path) {
+    LOG.trace("issuing delete for path to key: {}", path);
+    OFSPath ofsPath = new OFSPath(path);
+    String keyName = ofsPath.getKeyName();
+    try {
+      OzoneBucket bucket = getBucket(ofsPath, false);
+      incrementCounter(Statistic.OBJECTS_DELETED);
+      bucket.deleteKey(keyName);
+      return true;
+    } catch (IOException ioe) {
+      LOG.error("delete key failed " + ioe.getMessage());
+      return false;
+    }
+  }
+
+  /**
+   * Package-private helper function to reduce calls to getBucket().
+   * @param bucket Bucket to operate in.
+   * @param path Path to delete.
+   * @return true if operation succeeded, false upon IOException.
+   */
+  boolean deleteObject(OzoneBucket bucket, String path) {
+    LOG.trace("issuing delete for path to key: {}", path);
+    OFSPath ofsPath = new OFSPath(path);
+    String keyName = ofsPath.getKeyName();
+    try {
+      incrementCounter(Statistic.OBJECTS_DELETED);
+      bucket.deleteKey(keyName);
+      return true;
+    } catch (IOException ioe) {
+      LOG.error("delete key failed " + ioe.getMessage());
+      return false;
+    }
+  }
+
+  public FileStatusAdapter getFileStatus(String path, URI uri,
+      Path qualifiedPath, String userName)
+      throws IOException {
+    OFSPath ofsPath = new OFSPath(path);
+    String key = ofsPath.getKeyName();
+    try {
+      OzoneBucket bucket = getBucket(ofsPath, false);
+      incrementCounter(Statistic.OBJECTS_QUERY);
```

Comment:
should we move the counter incr before getBucket so that we can accurately count the ops even though getBucket may fail?

Reply:
Done in a9ef567103a13fd57d2c63197e7e356b447cafd7